Feature/calendar-and-event-creation

ScottBTR opened the pull request:
- Adding calendar creation
- Adding calendar event creation (no attendee creation yet)
- Added calendar deletion, 
  *  UWP doesn't have a deletion method outside of showing the calendar delete, not sure if I should try/can utilize this, for now have made it 'cancel' events.
  * Android Fully functiona. 
  * iOS needs further testing
- UI changes to support new features
- IsReadOnly being used to determine which calendar calendars I allow events to be added. but backend still needs to handle this gracefully.
- Need to look up validation of field, e.g. if I enter a location that's 10000 characters long, is that okay? will that work across platform? What exception should be thrown, and what other fields have limitations.

nickrandolph committed:
Added Calendar sample page with viewmodel
Added Calendar shared partial class
Added Calendar android partial class
Added Calendar (other platforms) partial class
Added IsSupported to all platforms
Added Calendar test for netstandard IsSupported

ScottBTR committed:
#1 - Allow Cheking and retrieving of permissions for android and iOS

Note:: For iOS there is no Calendar Read vs Write permissions, instead they have Calendar and Reminder permissions, which are quite different.
https://docs.microsoft.com/en-us/xamarin/ios/platform/eventkit

I've also noticed that although their is two seperate permissions for Calendar Read/Write for Android, when looking at the settings for the app after providing it with Read/Write Access it seems to set both as active, as there is no seperation in switches.

ScottBTR committed:
Allow Android to make a get calendars request.

NOTE:

Currently my CalendarObjects have a blocker where they cannot pull details from iOS or Android's provided libraries.

nickrandolph committed:
Fixing issue with disappearing properties on CalendarObject

nickrandolph committed:
Tidying up implementation code for Android

ScottBTR committed:
#2 Allow Xamarin to pull back a list of calendars for both iOS and Android

GetCalendarsAsync:
- Check/Request calendar read permissions
- Declare calendar instance/singleton
- Request List of iOS/Android platform specific calenders
- Cast to new shared List<DeviceCalendar> which obtains values for
  * Id - Platform specific android generally holds number ids whilst iOS uses guids
  * Name - Name of each calendar
  * IsReadOnly - Can this calendar be edited
- Currently the sample app only displays the name attribute, but we could indicate read only access and ids at a later state.

ScottBTR committed:
Amend changes set by PR for Issues - #1 #2 #3

ScottBTR committed:
Calendar Events able to be displayed for all or a specified calendar

ScottBTR committed:
Trial including start and end dates for event requests.

ScottBTR committed:
Allow Date & Time pickers to control event display (also works with specific calendars)

ScottBTR committed:
Allow Date & Time pickers to control event display (also works with specific calendars)

ScottBTR committed:
Provide code reference point

ScottBTR committed:
Added Checkboxes to enable and disable start date/time and end date/time, Also a modal!

Note - Might need help working out why my EventToCommandBehavior isn't working for my checkboxes.

They just use their property {getters/setters}, which I could just use but it was suggested that I use the EventToCommand approach where possible.

You can also get a modal to appear giving you extra information about an event when tapping on an item within the ListView.

ScottBTR committed:
Added Checkboxes to enable and disable start date/time and end date/time, Also a modal!

Note - Might need help working out why my EventToCommandBehavior isn't working for my checkboxes.

They just use their property {getters/setters}, which I could just use but it was suggested that I use the EventToCommand approach where possible.

You can also get a modal to appear giving you extra information about an event when tapping on an item within the ListView.

ScottBTR committed:
Added sort by datestart, changed listview to HasUnevenRows to make all event titles visible. altered date/time pickers to not use/need nullables. Including deleted events for the time being.

ScottBTR committed:
Added sort by datestart, changed listview to HasUnevenRows to make all event titles visible. altered date/time pickers to not use/need nullables. Including deleted events for the time being.

ScottBTR committed:
Added Github repository reference to class pulled from outside source.

ScottBTR committed:
Added new CalendarEventAddPage as a modal to enter data required to create an event

ScottBTR committed:
Added new CalendarEventAddPage as a modal to enter data required to create an event

ScottBTR committed:
Added new CalendarEventAddPage as a modal to enter data required to create an event

ScottBTR committed:
Add backend data call to be able to create new event when provided with a calendar id and required fields.

ScottBTR committed:
Add backend data call to be able to create new event when provided with a calendar id and required fields.

ScottBTR committed:
Clean up according to PR comments - Removal of not-needed converter EpocToDateTimeConverter

Removal of Reminders
Altered Recurrance variables to use the class RecurrancePattern (will require hooking up to database later on)

ScottBTR committed:
Clean up according to PR comments - Removal of not-needed converter EpocToDateTimeConverter

Removal of Reminders
Altered Recurrance variables to use the class RecurrancePattern (will require hooking up to database later on)

ScottBTR committed:
Remove Calendar Add Event feature [Android] - May be reimplemented at a later date.

ScottBTR committed:
Remove Calendar Add Event feature [Android] - May be reimplemented at a later date.

ScottBTR committed:
Made changes, addressing PR feedback

ScottBTR committed:
Addressing PR comment regarding reuse of static string.

ScottBTR committed:
Working out pulling back events for ios for any and specific calendars, currently NSDate to Epoch conversion still needs work..

ScottBTR committed:
Revert "Working out pulling back events for ios for any and specific calendars, currently NSDate to Epoch conversion still needs work.."

This reverts commit 973df83ef71b0d461d4e12b1a8146f8b42f989e9.

ScottBTR committed:
Revert "Revert "Working out pulling back events for ios for any and specific calendars, currently NSDate to Epoch conversion still needs work..""

This reverts commit 751f05d127c7a1e81d675c2977db3aeeb61e3e19.

ScottBTR committed:
iOS requesting events now works with date from/to, also sorts the resulting list by start date.

ScottBTR committed:
Alter Description to be iOS 'Notes' as 'Description' is actually the contents of the class.

Description: https://developer.apple.com/documentation/objectivec/nsobject/1418799-description

Notes:
https://developer.apple.com/documentation/eventkit/ekcalendaritem/1507507-notes

At a later date I should also probably use 'hasNotes' to return a more suitable message, or even as a lower level version of the event, that allows the user to determine whether they should request further data. (same with reminders, alarms and so forth)

ScottBTR committed:
don't need to await read access from viewmodel, each request already calls appropriate access

ScottBTR committed:
Added GetEventByIdAsycn --> Works at [Android] and [iOS] level currently.

ScottBTR committed:
[Android] Attendees can now be retrieved and rules are about half way. need to work out how the monthly and weekly list work from flutter.

ScottBTR committed:
Don't break when BYDAY= is the last parameter, Should add this on all parameters.

Must also first create the list before adding days to it.

ScottBTR committed:
Checks to make sure rules are non-nullable to prevent an unnessacary/crash from substringing.

ScottBTR committed:
UWP working with requesting read/write, pulling back calendares, pulling back events for calendars or a specific calendars, with time constraints. as well as event details.

NOTE: may need to work out a singleton pattern.

ScottBTR committed:
iOS Attendee Calendars

ScottBTR committed:
Don't try to build attendee list when there are none.

ScottBTR committed:
Allow Android to hold Deleted bool for Calendar and Event, so that users can decide whether to show those instances or not. Perhaps I need to make an android specific calendar extension hmmm

ScottBTR committed:
Add tests, and use default end date of 14 days away (there is a function that can override this setting)

ScottBTR committed:
iOS Hot.Reload and Hot.Restart

ScottBTR committed:
Addressing PR changes

ScottBTR committed:
Allow Uwp To see appointments, fix android trying to throw SQLException from string id.

ScottBTR committed:
Code cleanup to allow merge to be functional

ScottBTR committed:
Have Calendar creation (somewhat) working with creation, has strange permission issues that will need to be resolved.

ScottBTR committed:
Create Event requires string because iOS uses guids.

ScottBTR committed:
Address PR Comments

ScottBTR committed:
Addressing PR Comment in relation to not disposing of Cursor.

ScottBTR committed:
Allow AllDay to turn on/off time pickers, as if it's an all day event the time slots aren't required.

ScottBTR committed:
Add missing AllDay parameters for ios and uwp

ScottBTR committed:
Disable AddEvent button unless an event has a proper start/end date and an event name

ScottBTR committed:
Fixes to bad auto-merge plus add the ability to change the defaultstartingtimefromnow

ScottBTR committed:
Read-only Calendar API (#46)

* Calendar Read-Only API

* Return exceptions that match test case System.NotImplementedException --> Xamarin.Essentials.NotImplementedInReferenceAssemblyException (#45)

ScottBTR committed:
Replicate behaviour accross platforms for getevents when passing in an incorrect calendar id.

ScottBTR committed:
Added word to make id more descriptive for comment.

ScottBTR committed:
Undo space removal.

Redth committed:
Add Share Source Position (#1002)

* Add SourceScreenPosition to Share*

* Rename base type to be more specific

* Add docs for Share changes

* Add SourceScreenPosition to sample

* Rename source property

* Fixed docs from rename

* Fix docs

* Another docs fix

* Make abstract

ScottBTR committed:
Adding On Delete Button

mrlacey committed:
Rename GetCompliment to GetComplementary (#1016)

For #1012

ScottBTR committed:
Allow Creation of Calendars to work cross platform. (iOS still needs testing but blocked by apple update!)

jamesmontemagno committed:
Update Xamarin.Essentials.csproj

ScottBTR committed:
Replace fake ios calendar event delete code with throw not implemented exception.

ScottBTR committed:
Revert xamarin essential project alterations.

ScottBTR committed:
Cross-platform deletion

Note - UWP doesn't have a deletion method outside of showing the calendar delete, not sure if I should try/can utilize this, for now have made it 'cancel' events.

Fully functional on Android.

iOS needs testing but getting blocked by odd issues at present.

nickrandolph commented on Samples/Samples.iOS/Info.plist:
```diff
@@ -55,8 +55,10 @@
-  <string>Access to view and create calendar events</string>
+    <key>NSCalendarsUsageDescription</key>
+    <string>Access to view and create calendar events</string>
+    <key>NSRemindersUsageDescription</key>
```
Just confirming we need this for writing basic calendar events?

ScottBTR replied:
Ahh I suppose we wouldn't need it yet, I believe, happy to remove it at this stage if we're not planning on tackling reminders in the near future/getting a release out inbetween

nickrandolph replied:
Yes, we should only include absolute minimum permissions we need for what we're doing

ScottBTR replied:
Altered accordingly

ScottBTR committed:
[UWP] - Add/Remove Attendees for Events now functional

ScottBTR committed:
[Android] Add/Remove attendee from event working, if duplicate attendees exist all will be removed at once.

jamesmontemagno committed:
Fix issue with checking mainthread to early on UWP. (#1021)

jamesmontemagno committed:
Update FileSystem.xml (#1015)

jamesmontemagno committed:
Cleanup permissions check for Geolocation (#1022)

* Cleanup permissions check for Geolocation

* Return false if feature or name is null.

* Ensure tests run properly on android

jamesmontemagno committed:
Cleanup equals in IEquatable so no boxing occurs. (#1019)

ScottBTR committed:
Improve user navigation + add remove attendee for iOS

jamesmontemagno committed:
[WIP] GH-923 Implement AppTheme for iOS/tvOS/Android/UWP (Need Tizen Still) (#927)

* Implement AppTheme for iOS/tvOS/Android/UWP (Need Tizen Still)

* Added docs

* Added sample

* Added a test

ScottBTR committed:
Updates - UWP/workflow

UWP - now removes one attendee when multiple of the same attendee are present
EventPage - now reloads upon entry, which means if an attendee was added when you go back to it, this will now be reflected immediately
MainPage - no longer reloads the events manually upon entry, as this is triggered from reloading the calendars which was resulting in doubled entries in the event list.

ScottBTR committed:
ios: Update Add Attendee with details about ios limitations, fix add calendar creation for iOS

ScottBTR committed:
Android/ios/uwp Calendar creation and Event creation/update/delete now fully functional

ScottBTR committed:
[iOS]: Add Commented code that shows the idea behind removing attendees

ScottBTR committed:
Fix formatting of event page and add calendar title when editing an event.

ScottBTR committed:
Split Create and Update into two functions to create separation of concern.

ScottBTR committed:
Attendance required status can now be set and visible by color

ScottBTR committed:
Recurrence Rule Changes

jamesmontemagno committed:
Fixes #990 - only pull in utils when needed (#1033)

* Fixes #990

* Update azure-pipelines.yml

* Add timeout for running android tests.

jamesmontemagno committed:
Update azure-pipelines.yml

ScottBTR committed:
Added a basis for Rules/Reminder retrieval

Redth committed:
Update azure-pipelines.yml for Azure Pipelines

Redth committed:
Update azure-pipelines.yml for Azure Pipelines

Redth committed:
Update azure-pipelines.yml for Azure Pipelines

Redth committed:
Update azure-pipelines.yml for Azure Pipelines

Redth committed:
Update azure-pipelines.yml for Azure Pipelines

Commit:
GH-969 Clipboard Changed (#720)

* Use SharedPreferences.Editor.Apply when editing to force disk write to be asynchronous (#637)

* GH-182 Color/Point/Rect/Size Extension Converters (#651)

* Add extension helpers for iOS and Android

* Add UWP colors and adjsut android colors.

* Add UWP Point/Rect/Size

* Rename and optimize!

* Add tests and color helpers!

* Fix unit test and add WithAlpha

* Cleanup iOS as it stores it from 0-1

* Tests are all green!

* Add generate-docs cmd :)

* Add docs config to all :)

* Additional clenaup for docs

* Added some docs

* Updated the mdoc target and regenerated the docs

* Allow for code reuse on macOS (#665)

Even though macOS is not yet officially supported, it is nice to reuse this code. macOS is exactly the same, except for this single property not supported.

* GH-196 Browser Customization (#646)

* Implemented GH-196

* Apply suggestions from code review: Naming

Renamed incorrectly spelled variables

Co-Authored-By: Mrnikbobjeff <schillinik@yahoo.de>

* Implemented GH-196

* Browser update work as requested. Using System.Drawing.Color now

* Adopted more review changes. Type forwarded.

* Review changes

* Made equals operator null safe, removed constructor

* Add documentation and simplify the API!

* Update viewmodel

* Back to 7.2

* GH-676 Require To Check GPS Location is Fake Or Not In GeoLocation (#677)

* Added  bool IsFromMockProvider in Location class

* Checkin

* csproj original

* Revert "csproj original"

This reverts commit 6302b95782006288862f0ce11100c81936baa7fb.

* Revert "Checkin"

This reverts commit 0881e4f12c36caecd08b3b9bb8757d8bd1025502.

* Revert "Added  bool IsFromMockProvider in Location class"

This reverts commit e56d9406cf0a92d24b04607afe81aef23aca696c.

* Revert the Xamarin.Essentials.csproj and Samples.Android.csproj to original;

Adjust the property IsFromMockProvider in Location.shared and LocationExetensions.

* Edited sample to show if the location is from moking provider.

* Update docs

* Fixes #694 (#699)

* Update the mdoc to use the new minimum for VS2019

* GH-126: Finish Shake Detector API (#693)

* Update CONTRIBUTING.md (#692)

* GH-126 Detect Shake API (#666)

* Added shaken support

* name vhange

* check-in

* doc update

* Implements the Shake API inside Accelerometer Class;
Change sample to AccelorometerViewModel

* Fix the sample project

* Update the docs

* Added Queue mechanism based off seismic with tests as well. Something is not  right yet with calulating isaccelerating.

* Finalize shake detection!!!

* Re-generated the docs.

* Fix the VM after merge

* GH-704 Handle duplicate item in keychain (#705)

* Update CONTRIBUTING.md (#692)

* #704 if we get a duplicate item try to remove and then re-add if possible.

* Initial commit Clipboard Change

* Implemented Clipboard listeners on iOS, tested on android and uwp and fixed other issues related to Visual Studio for Mac and Essentials support.

* Initial commit Clipboard Change

* Implemented Clipboard listeners on iOS, tested on android and uwp and fixed other issues related to Visual Studio for Mac and Essentials support.

* Naming fix

* UI improvements

* Docs update

* docs update part 2

* docs tool breaking fix

* reverted change to csproj

Redth committed:
Update azure-pipelines.yml for Azure Pipelines

Redth committed:
Update azure-pipelines.yml for Azure Pipelines

Redth committed:
Disable the pesky comments?

Redth committed:
Update azure-pipelines.yml

Redth committed:
Permissions API (#949)

* Initial (partial) effort

* Get rid of old nuget feed

* New permission API

* Bump tests framework

* Create sample for permissions API

* Add some iOS info.plist usages

* Show an error on permission exception in sample

* Clean up iOS permission implementation

* Rename platform base permission impl

So that all platforms are the same and inheritance looks the same on each

* Simplify static permissions api impl

* Make more permission inheritance consistent

* Preserve no longer needed on android

Since we are instantiating the permission types properly now, the linker is happy.

* Cleanup API with C# 8 and optimize for Android 10

* Cleanup tests and samples

jamesmontemagno committed:
First attempt at cellular data on iOS fix (#1034)

nickrandolph committed:
Fixing up permissions usage

ScottBTR committed:
Bring accross UI code for displaying Recurrence Rules

ScottBTR committed:
Using new UWP recurrence rule model

(with a readonly version existing to either allow us to view rules we do not process the power to create, or to revert back to at a later stage)

the iOS and Android is commented out at this stage as they will most likely require reworks once we have a clear vision going forward.

the selection of days of the week works for selection but not initilization will be looking to convert to a list of checkboxes tomorrow if I am unable to solve it within the first hour.

ScottBTR committed:
Converted CollectionView of Observable collection to use Switches

Altered class name TestDumbReference --> DayOfTheWeekSwitch to match UI elements/make logical sense

ScottBTR committed:
Code Tidy up

ScottBTR committed:
Changes to be more consistent with our OSs

- Allow Intervals to be set up to 99 instead of limiting to a drop down up to 10
- Allow an an end point to be set by number of iterations rather than relying on end date.

Improve UI for the new fields/changes

ScottBTR committed:
Updating to UI and initial set up for iOS

Starting work on iOS to be able to read/update recurrence rules following the UI in place

Altering end point UI to match flutters drop down menu control system as well as adding validation that makes sure values exist when looking for a date/occurrences before allowing creation/update of the event to take place

ScottBTR committed:
Updates to UI and cross-platform attendee types implemented

ScottBTR committed:
Android fixes, plus timezone fixes for android + iOS

ScottBTR committed:
Updates

- Recurrence Rules moved to converter for android/ios to work
- Start date now consistent in design when all day event
- Testing out ToLocalTimeConverter to help with date consistency issues, currently only partially working with android
- Getting event INSTANCES in trial with android
- Added url validator
- In process of adding attendee email validator
- In process of deletion by event instance instead of only by entire instance
- UWP can kinda delete bookings now, but has to be done through windows UI
- In process of adding attendees for iOS

ScottBTR committed:
Get TimeZones && Recurrence Rule Frequencies working more reliably accross all platforms.

ScottBTR committed:
Meeting PR Feedback

ScottBTR committed:
Changes List below:

- RecurrenceRuleReadOnly removed at this stage.
- enums with .None or .NotSet have been altered to no longer include those values and instead values that require a default non-set value have been made nullables
- iOS recurrence rules:
  * Now properly deals with .Last (-1 not 4)
  * Moved Recurrence rule setters into separate function to reduce duplicated code between create and update
  * Saving recurring events now effects all instances of events instead of only the one being edited
- UWP will now automatically add http:// when not included to reduce pain for developers/users as uwp requires a URL instead of string value
- DayIterationOffsetPosition --> WeekOfMonth for readability
  * IterationOffset Enum (should be name changed too) is now set up by platform to allow for easy straight conversions.
- MonthOfTheYear --> MonthOfYear

ScottBTR committed:
Set up for some expected pass test cases, will be refactored at a later date.

mkieres commented on Samples/Samples/View/CalendarAddPage.xaml:
```diff
@@ -0,0 +1,45 @@
+        <Grid.RowDefinitions>
+          <RowDefinition/>
+          <RowDefinition/>
+          <RowDefinition/>
```
Do we need that many rows? It looks like we're using just 1

ScottBTR replied:
Yeah wow, not sure why that's been left in, should I fix this in the latest branch or this one?

mkieres commented on Samples/Samples/View/CalendarAddPage.xaml.cs:
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.ObjectModel;
```
We could probably clean up some of the using statements that are not used - very low priority to fix

ScottBTR replied:
Fair point 👍

mkieres commented on Samples/Samples/View/CalendarEventPage.xaml:
```diff
@@ -17,11 +17,26 @@
                    HorizontalTextAlignment="Center"/>
-          <Grid>
+                <Button Grid.Column="1"
+                      Text="Delete"
```
Formatting: Fix attribute alignment in the file

ScottBTR replied:
I believe this is fixed/altered in latest branch/PR

mkieres commented on Samples/Samples/View/CalendarEventPage.xaml.cs:
```diff
@@ -11,5 +11,20 @@ public CalendarEventPage()
+            if (!(EventId.Text is string eventId) || string.IsNullOrEmpty(eventId.ToString()))
+                return;
+
+            var calendarEvent = await Calendar.GetEventByIdAsync(eventId.ToString());
```
Is there a need to call `ToString()` on the `eventId` here? I might be wrong but the logic would indicate that the `eventId` will be a string anyhow

ScottBTR replied:
Yep, pointless conversion

mkieres commented on Samples/Samples/View/CalendarPage.xaml.cs:
```diff
@@ -14,6 +14,25 @@ public CalendarPage()
+        {
+            var modal = new CalendarEventAddPage();
+
+            if (!(SelectedCalendar.SelectedItem is DeviceCalendar tst) || string.IsNullOrEmpty(tst.Id))
```
The `tst` name of the device calendar is a bit odd. Can we rename this variable to something more accurate?

ScottBTR replied:
Yep, my bad I'm sure that's not the only occurrence, I need to remember that once I've proven that my theory works I've gotta go back over and fix the temporary names...

for (var x = 0; x < TimesUsingBadVariableName; x++)
{
    System.Debug.Log($"You're correct for the {x}th time");
}

mkieres commented on Samples/Samples/ViewModel/CalendarAddViewModel.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        public CalendarAddViewModel()
+        {
+            CreateCalendar = new Command(CreateCalendarCommand);
```
You could move this code our of the constructor, like so:

`public ICommand CreateCalendar { get; } = new Command(CreateCalendarCommand);` (I think :) )

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -82,11 +97,11 @@ static async Task<IEnumerable<DeviceEvent>> PlatformGetEventsAsync(string calend
                 }
             }
-            if (events.Count == 0)
+            if (events.Count == 0 && !string.IsNullOrEmpty(calendarId))
```
I tend to use `Any()` extension method instead of `Count`, as it's a bit easier to read - not too fussed about this.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -209,5 +224,75 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(string eventId)
+            await Permissions.RequireAsync(PermissionType.CalendarWrite);
+
+            var calendarUri = CalendarContract.Calendars.ContentUri;
+            var cursor = Platform.AppContext.ApplicationContext.ContentResolver;
```
Just double checking if the name of this variable `cursor` is correct? I looks a bit off but I'm not too familiar with ContentResolver so maybe the name it's adequat, in which case just ignore my comment

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -209,5 +224,75 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(string eventId)
+            }
+            catch (NullReferenceException ex)
+            {
+                throw ex;
```
What's the point of catching and re-throwing an exception here? Is this something that Essential team is doing?

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -209,5 +224,75 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(string eventId)
+            await Permissions.RequireAsync(PermissionType.CalendarWrite);
+
+            var result = 0;
+            if (string.IsNullOrEmpty(newEvent.CalendarId))
```
Should we be a bit more defensive here and make sure that the `newEvent` is not null?

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -209,5 +224,75 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(string eventId)
+            eventValues.Put(CalendarContract.Events.InterfaceConsts.EventLocation, newEvent.Location);
+            eventValues.Put(CalendarContract.Events.InterfaceConsts.AllDay, newEvent.AllDay);
+            eventValues.Put(CalendarContract.Events.InterfaceConsts.Dtstart, newEvent.StartDate.ToUnixTimeMilliseconds().ToString());
+            eventValues.Put(CalendarContract.Events.InterfaceConsts.Dtend, newEvent.EndDate.HasValue ? newEvent.EndDate.Value.ToUnixTimeMilliseconds().ToString() : newEvent.StartDate.AddDays(1).ToString());
```
It looks pretty arbitrary that we're adding a day if the end date hasn't been provided. Couple of questions:

1. Should we be even providing an end date if it's not provided?
2. Why are we adding exactly 1 day?

If we need to provide a value and it will be an arbitrary number, I feel that we might want to leave a comment why about it.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -209,5 +224,75 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(string eventId)
+            eventValues.Put(CalendarContract.Events.InterfaceConsts.Dtstart, newEvent.StartDate.ToUnixTimeMilliseconds().ToString());
+            eventValues.Put(CalendarContract.Events.InterfaceConsts.Dtend, newEvent.EndDate.HasValue ? newEvent.EndDate.Value.ToUnixTimeMilliseconds().ToString() : newEvent.StartDate.AddDays(1).ToString());
+            eventValues.Put(CalendarContract.Events.InterfaceConsts.EventTimezone, TimeZoneInfo.Local.Id);
+            eventValues.Put(CalendarContract.Events.InterfaceConsts.Deleted, 0);
```
Is this necessary? It seem a bit weird that you need to set the `Deleted` to false flag when **creating** an event (because who would want to create a deleted event?). Although I wouldn't be surprised if Android APIs would require it.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -209,5 +224,75 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(string eventId)
+            try
+            {
+                var resultUri = Platform.AppContext.ApplicationContext.ContentResolver.Insert(eventUri, eventValues);
+                result = Convert.ToInt32(resultUri.LastPathSegment);
```
I would probably try to use `int.TryParse` instead and that way avoid potential exceptions, unless throwin an exception is expected.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.android.cs:
```diff
@@ -209,5 +224,75 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(string eventId)
+
+        static async Task<bool> PlatformDeleteCalendarEventById(string eventId, string calendarId)
+        {
+            await Permissions.RequireAsync(PermissionType.CalendarWrite);
```
Should we be checking here also read permissions, as we're retrieveing the event? Or if you have write permissions it also let's you read?

ScottBTR replied:
I believe you require write permissions to be able to read, but will validate this and get back to you.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.ios.cs:
```diff
@@ -108,5 +109,60 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<EKParti
+        {
+            await Permissions.RequireAsync(PermissionType.CalendarWrite);
+
+            var evnt = EKEvent.FromStore(CalendarRequest.Instance);
```
Similar suggestions/comments as in the Android implementation.

1. Worth considering a bit more defensive code to check if `newEvent` is null or not
2. Checking if `EndDate` is required or adding a comment about the +1 day to `EndDate`

mkieres commented on Xamarin.Essentials/Calendar/Calendar.ios.cs:
```diff
@@ -108,5 +109,60 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<EKParti
+
+            var calendarEvent = CalendarRequest.Instance.GetCalendarItem(eventId) as EKEvent;
+
+            if (calendarEvent.Calendar.CalendarIdentifier != calendarId)
```
I would try to follow a general rule of thumb of using null propagation when casting to specific types, as the casting might fail and the variable will end up being a null and you will get a null reference exception.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.ios.cs:
```diff
@@ -108,5 +109,60 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<EKParti
+        {
+            await Permissions.RequireAsync(PermissionType.CalendarWrite);
+
+            var calendar = EKCalendar.Create(EKEntityType.Event, CalendarRequest.Instance);
```
So to create a calendar you need to create an event? It might be correct but just wanted to double check.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.uwp.cs:
```diff
@@ -134,5 +137,86 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<Appoint
+                Duration = newEvent.EndDate.HasValue ? newEvent.EndDate.Value - newEvent.StartDate : TimeSpan.FromDays(1),
+                AllDay = newEvent.AllDay
+            };
+            try
```
Formatting: Add spacing

mkieres commented on Xamarin.Essentials/Calendar/Calendar.uwp.cs:
```diff
@@ -134,5 +137,86 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<Appoint
+            }
+            catch (NullReferenceException ex)
+            {
+                throw ex;
```
Same question as in my previous comment - what's the purpose of re-throwing an exception?

mkieres commented on Xamarin.Essentials/Calendar/Calendar.uwp.cs:
```diff
@@ -134,5 +137,86 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<Appoint
+
+            var instance = await CalendarRequest.GetInstanceAsync();
+
+            var app = new Appointment()
```
Not too fussed about this but I would try to avoid abbreviations in code and use full names instead

mkieres commented on Xamarin.Essentials/Calendar/Calendar.uwp.cs:
```diff
@@ -134,5 +137,86 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<Appoint
+
+            try
+            {
+                // var appointmentStore = await AppointmentManager.RequestStoreAsync(AppointmentStoreAccessType.AppCalendarsReadWrite);
```
Remove commented out code

mkieres commented on Xamarin.Essentials/Calendar/Calendar.uwp.cs:
```diff
@@ -134,5 +137,86 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<Appoint
+
+            var cal = await instance.GetAppointmentCalendarAsync(calendarId);
+            var app = await instance.GetAppointmentAsync(eventId);
+            app.IsCanceledMeeting = true;
```
Should we leave a comment here about inability to actually delete an appointment and therefore we're canceling it?

ScottBTR replied:
Have actually decided to use uwp's UI deletion in the latest branch/PR, unfortunately no way to do it programatically but it is functional

mkieres commented on Xamarin.Essentials/Calendar/Calendar.uwp.cs:
```diff
@@ -134,5 +137,86 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<Appoint
+            app.IsCanceledMeeting = true;
+            await cal.SaveAppointmentAsync(app);
+
+            if (app.IsCanceledMeeting)
```
I'm not sure if this is necessary? Can the `SaveAppointmentAsync` revert the `IsCanceledMeeting` flag to false or it will throw an exception if it won't be able to cancel the appointment?

ScottBTR replied:
No longer needs to call SaveAppointmentAsync using the new deletion method mentioned above.

mkieres commented on Xamarin.Essentials/Calendar/Calendar.uwp.cs:
```diff
@@ -134,5 +137,86 @@ static IEnumerable<DeviceEventAttendee> GetAttendeesForEvent(IEnumerable<Appoint
+
+        static async Task<bool> PlatformDeleteCalendarEventById(string eventId, string calendarId)
+        {
+            await Permissions.RequireAsync(PermissionType.CalendarWrite);
```
A general note to be aware that there have been permissions changes in the code base (see 1.4 pre release notes https://docs.microsoft.com/en-us/xamarin/essentials/release-notes/1.4), which might impact how this code is executed.

ScottBTR replied:
Ahh yes, we've applied these permission changes to the read-only version but both the naming class naming convention and permission changes have not been migrated over.

mkieres commented on Xamarin.Essentials/Permissions/Permissions.ios.cs:
```diff
@@ -137,5 +157,23 @@ static Task<PermissionStatus> RequestCalendarAsync()
+                EKEntityType.Reminder,
+                (bool granted, NSError e) =>
+                {
+                    if (granted)
```
Something not right here - empty clause. You might want to revert the condition and remove the empty clasue

ScottBTR replied:
Yep it should probably follow the same format as the EKEventType.Event instance above it:
(bool granted, NSError e) =>
{
    tcs.SetResult(granted ? PermissionStatus.Granted : PermissionStatus.Denied);
});

mkieres commented on Xamarin.Essentials/Types/PlatformExtensions/CalendarExtensions.shared.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Xamarin.Essentials
+{
+    public static partial class CalendarExtensions
```
I woder what's their naming convention for these. We generally follow a rule of naming the class file after the extended type. In this case it would be `TimeSpanExtensions`.

I also wonder if dropping `Round` in the name of the method is sort of redundant. I also like better when the conversion of a value to different value is done with a method named `To<Something>` (e.g. `ToString()`).

mkieres reviewed (commented):
See comments

ScottBTR committed:
PR feedback alterations.

ScottBTR committed:
Feature/bridge naming convention changes (#69)

ScottBTR committed:
Feature/merge readonly calendar (#70)
